src/qutip_qip/circuit/circuitsimulator.py

```diff
@@ -559,38 +453,43 @@ def _decimal_to_binary(decimal, length):
             binary = [int(s) for s in "{0:#b}".format(decimal)[2:]]
             return [0] * (length - len(binary)) + binary
 
-        op = self.ops[self.op_index]
-        if isinstance(op, Measurement):
-            self._apply_measurement(op)
-        elif isinstance(op, tuple):
-            operation, U = op
-            apply_gate = all(
-                [
-                    self.cbits[cbit_index] == control_value
-                    for cbit_index, control_value in zip(
-                        operation.classical_controls,
-                        _decimal_to_binary(
-                            operation.classical_control_value,
-                            len(operation.classical_controls),
-                        ),
-                    )
-                ]
+        def _check_classical_control_value(operation, cbits):
+            """Check if the gate should be executed, depending on the current value of classical bits."""
+            matched = np.empty(len(operation.classical_controls), dtype=bool)
+            cbits_conditions = _decimal_to_binary(
+                operation.classical_control_value,
+                len(operation.classical_controls),
             )
-            if apply_gate:
-                if self.precompute_unitary:
-                    U = expand_operator(
-                        U,
-                        dims=self.qc.dims,
-                        targets=operation.get_all_qubits(),
-                    )
-                self._evolve_state(U)
-        else:
-            self._evolve_state(op)
+            for i in range(len(operation.classical_controls)):
+                cbit_index = operation.classical_controls[i]
+                control_value = cbits_conditions[i]
+                matched[i] = cbits[cbit_index] == control_value
+            return all(matched)
 
-        self.op_index += 1
-        return self.state
+        op = self.qc.gates[self._op_index]
```

Comment:
```suggestion
        op = self._qc.gates[self._op_index]
```
Not sure if this is definitely better, but maybe nice to avoid the property lookup.